### ISSUE TITLE
Improve component with extra registers for multiple device types

### ISSUE THREAD

**Issue by htvekov** [cfeenstra1024/kamstrup-multical-hardware#1]
Hi' @cfeenstra1024

Still waiting for the ESPHome PR to be merged before testing 🙂
I'm currently using ESPHome with my own 'tweaked' version of @cenobitedk Kamstrup custom component. Both for my district heating meter (Multical 602) as well as my water meter (flowIQ 2200).

I would suggest that you add the High Resolution register(s) to your component as well. For the x03 meter series it's a straight forward process. For the x02 series it's a bit more complicated as the HiRes register is a delta value. I've also added a calculated yearly cooling average sensor. Very useful in countries where the utility companies enforce 'penalty' tariffs for customers not cooling the water sufficiently. Reading the Unit Of Measurement from the actual device would also be a nice feature (not sure if you've already done that ?)

You could take a peek at my [fork](https://github.com/htvekov/esphome_multical402) if you're interested 🙂

**Comment by cfeenstra1024** [comment]
Hi @htvekov, 

Nice, you have different meters than I have. I only have a Multical 403. The component I created has native support for some of the most common registerers like 'Heat Energy' and 'Power'. However, you can add any number of registers yourself using the config file. See an example below that reads the high resolution register you've mentioned. I did not really want to add all possible registers in the component, but instead allow the (advanced) user to enter the register IDs themselves. 

Once the PR is merged, I am open to extending the list of sensors. But I'm really keen to have this component first in ESPHome, before I disturb reviewers with more commits ;)

I also thought about using the unit read from the device, and set that to the sensor. However, this seems not to be possible / allowed. After I'd created this, I received [some comments](https://github.com/esphome/esphome/pull/4200#pullrequestreview-1347750261) about that in the PR.

A documentation PR is also pending, explaining the config options. As long as this PR is open, you can check it out [here](https://deploy-preview-3513--esphome.netlify.app/components/sensor/kamstrup_kmp).

```yaml
uart:
  baud_rate: 1200
  stop_bits: 2
  tx_pin: GPIO15
  rx_pin: GPIO13

sensor:
  - platform: kamstrup_kmp
    heat_energy:
      name: 'Heat Energy'
    power:
      name: 'Heat Power'
    temp_diff:
      name: 'Heat Temperature Difference'
    flow:
      name: 'Heat Flow'
    temp1:
      name: 'Temp 1'
    temp2:
      name: 'Temp 2'
    volume:
      name: 'Volume'
    custom:
      - name: 'High Resolution Heat Energy'
        command: 0x010A
        unit_of_measurement: 'Wh'
      - name: 'Some other custom sensor (Power)'
        command: 0x0050
```

**Comment by htvekov** [comment]
I understand your wish to get the initial PR done before adding functionality 😉

It makes perfect sense to keep the quantity of default sensors low. I like the idea of the possible custom sensors.
Perhaps later you could add a class (district heating, water, electricity) to define which default sensors should be present ?
The registers needed for the FlowIQ 2200 water meter are almost all different. And you might come a bit short with only five custom sensors.  

Shame about the UOM. Seems that there is no easy way to update that dynamically. Shame, as the UOM acually can be read from the meter.

**Comment by cfeenstra1024** [comment]
I'll think about how I can add the different registers in a way the code stays maintainable and the configuration in the yaml file is not too complex.
The custom sensor supports any number of registers (it is a list). So no limitation to 5 any more.

**Comment by htvekov** [comment]
For the Kamstrup FlowIQ 2200 meter following registers imo are the most relevant. Flow and volume registers are identical with the 40x/60x meters.

As this meter can detect both leak, burst, dry and tamper I'm currently searching for relevant Multical 21/FlowIQ 2x00 meter register info. So far I've not been able to find any info that confirms that these actually are the right registers for the info bit register and the reverse flow. The three lsb's should represent the actual event. Remaining bits should be a 'coarse' time indication for the duration. I'm not really keen to do a real life test on the information register - simulating a leak by running the tap for 24 hours 😉

The meter should also be able to report both ambient- and inlet water average temperature. I'm currently trying to identify these registers manually. So far no luck. But I've identified other registers such as max. flow pr. day etc.

```
0x00F3, // 0243 - V1 reverse flow				UOM 0x0028 (40 dec.): 'm³'
0x004A, // 0074 - Current water flow	 		UOM 0x0029 (41 dec.): 'l/h'
0x0044, // 0068 - Volume register V1			UOM 0x0028 (40 dec.): 'm³'
0x0063, // 0099 - Information					UOM 0x0033 (51 dec.): 'No unit'
0x03EC  // 1004 - Operating hour counter		UOM 0x002E (46 dec.): 'h'
```

**Comment by htvekov** [comment]
I would still add the district heating HiRes Energy register though. This is highly usable to get more detailed resolution on the hourly readings e.g. in HA's Energy dashboard. Confirmed working on the new meter series. Not really sure what register the 402 meter is using. Should be 0x009B, but not confirmed. 

```
0x010A, // 266 - Heat energy - Multical 403/603 High resolution (1 Wh)
0x009B  // 155 - Heat energy - Multical 602 High resolution (0.01 kWh)
```

![image](https://github.com/cfeenstra1024/kamstrup-multical-hardware/assets/45469813/e16cc652-f0fd-4893-8a1f-856ffc4501a1)

**Comment by cfeenstra1024** [comment]
What is the resolution of your 'Heat Energy' register? I get the value in GJ with 3 decimals. That is more than enough to entertain the energy dashboard. I did convert GJ to m3 in HA, as it did not support GJ in the energy dashboard when I started this project (not sure if it does now).

I do not really understand the high res register in Wh on the 403. The value it provides does not correspond with the GJ value. 
I get 6655807 Wh when reading the value from the 403. That corresponds with 23.961 GJ. However, the actual GJ value on the meter is close to 60. 

![image](https://github.com/cfeenstra1024/kamstrup-multical-hardware/assets/73584137/6b73ec1a-b1fb-4cfb-913f-a58ee7804726)

**Comment by htvekov** [comment]
Well, it seems that the HiRes register values returned differs depending on utility company/country configuration.

If you check the cross referenced link further up this thread, two users have confirmed that the HiRes register is working as expected.

A Danish guy with a 403 meter (configured to return energy consumption in kWh). Here the HiRes register is also reported in kWh but with 3 decimals.
Also Dutch guy using a 603 meter with consumption i GJ for the 'normal' energy register. Here the HiRes register is reported in MWh with 5 decimals. 

It seems that your HiRes register is returning a 'delta' value just a I get with my 602 meter. I guess that I've triggered and started my HiRes register at some point when I first attached the IR optical head - And it has been running ever since (4 months or so).
Upon ESPHome boot I deduct the HiRes register from the 'normal' energy register and store that difference as a constant. I add that constant to the HiRes register and return that value as my 'real' HiRes register.

My HiRes register is returned in MWh with 5 decimals. My 'normal' energy register is returned as well in MWh but with three decimals only (1 kWh resolution)

Note that ESPHome has issues calculating with such large floats. My calculations are sometimes off on the last decimal as 32bit float can't handle my register values. AFAIK it's not possible to change the float type in ESPHome.

![image](https://github.com/cfeenstra1024/kamstrup-multical-hardware/assets/45469813/45450c6c-8053-4df5-a5c2-2e0692267850)
